### ISSUE TITLE
[Suggestion] Save Transforms to Scene

### ISSUE THREAD

**Issue by DillonSteyl** [HungryProton/scatter#111]
I'd like to suggest we add the _option_ to save/load the mesh transforms from the scene file directly. I wanted to gauge interest in this feature before starting work on a pull request.

## Use Case
Reading pre-written transforms from disk can be more efficient than re-running the modifier stack every time the scene is loaded - although it does result in larger scene files. It also means that you can ensure the meshes are already placed when the node is added to the scene tree, rather than writing some custom loading functionality based on the `build_completed` signals of the modifier stacks.

In my case, I'm already using Godot's `ResourceLoader` class to load my levels, and I don't want to have to implement additional loading functionality to detect when the scatter nodes are finished building.

## Implementation Details
I have a very basic proof of concept up and running which caches each modifier stack's `TransformList` and saves it to the scene file, e.g.
```tscn
[sub_resource type="Resource" id="Resource_k7agn"]
resource_local_to_scene = true
script = ExtResource("27_1368b")
stack = Array[Resource]([SubResource("Resource_f0y5b"), SubResource("Resource_5g2ce"), SubResource("Resource_wtale"), SubResource("Resource_woo6r")])
cached_transform_list = SubResource("Resource_hw61j")
```
The `cached_transform_list` contains the list of transforms for this particular ModifierStack.
In the stack's `_ready()` function, you can simply check if that list exists, and if so, call the `_on_transforms_ready` function directly to place the meshes without re-running all the modifiers.

Some additional thoughts:
- If you have many objects, this can make the scene file quite large. It's not too big a problem when using binary `.scn` files, but when using `.tscn` files it can make version control very slow & tedious. Perhaps we could implement something similar to lightmap baking - i.e. save the transforms for all `scatter` nodes to a separate binary file and load them from there.
- It's worth having this option be specific to each scatter node. For one scene, you may want some scatter nodes to load asynchronously and other nodes to load immediately.
- For maximum efficiency it's probably worth caching the transforms as a [PackedFloat32Array](https://docs.godotengine.org/en/stable/classes/class_packedfloat32array.html#class-packedfloat32array) and then using the multimesh [buffer](https://docs.godotengine.org/en/stable/classes/class_multimesh.html#class-multimesh-property-buffer) property.

**Comment by HungryProton** [comment]
It can be done and would probably be a better option than the workarounds we already have in v4 today.

In the old v3, the internal scatter nodes were visible in the tree by default (the multimesh and duplicates). Which means they were also stored locally in the scene files, and as you mentioned, these files got quite big (But, you could save the result as is and not regenerate it every time).

So in v4 I made them internal by default to fix the large file issue and declutter the scene tree. But you can revert that behavior on individual Scatter nodes by ~~enabling~~ setting these two options:
+ `General / Show ouput in tree` - Enabled
+ `Performance / Force rebuild on load` - Disabled

Now, if you want the above without showing internals nodes in the tree, or without cluttering your scene file, we can implement your proposal.

But keep in mind that even with the transforms list ready, if you're not using instancing, adding the nodes to the scene tree is also a heavy operation depending on how many nodes there are. For now that's not an issue because it's doing it in a single frame, but I wonder If this step shouldn't be spread out to avoid lag spikes. 

If you're using instancing, though, no issues on that side.

**Comment by DillonSteyl** [comment]
Ah, awesome! These options weren't available in the version I had cloned, but they serve this purpose well. Although, to be clear, I'd keep "force rebuild on load" disabled in order to use the placements directly from the scene tree.

I'm not too fussed about the multi-meshes being displayed in the scene tree, so perhaps this proposal is not required. However, I am still curious about the idea of baking the transforms into an external binary file to keep the scene files minimal - this would be especially useful for things like grass, where you might have thousands of transforms being stored. But I think that probably belongs in a separate ticket. What do you think @HungryProton - is it worth considering this as a feature?

For now, I'm happy to just save my levels as binary `.scn` files to keep the size down.

**Comment by HungryProton** [comment]
My bad, I edited the comment above with the right values. 
I'm keeping the issue open, it could still be useful for version control purposes. I tend to keep my scenes in text format, but I definitely don't need to keep track of every single transforms, so moving them in their own resource file should work better.

It's not an immediate priority, but still something to consider.

**Comment by DillonSteyl** [comment]
Okay, I'm happy to take an initial look into getting that working if you'd like.

Do you think that external file would also be on a per-scatter-node basis, or would you somehow want to combine all of the scatter transforms for one scene into a single file?

On the one hand, it could get annoying saving every scatter node's transforms into an external file. On the other, its nice to have that level of granular control - some of the transforms may not need to be exported into another file.

**Comment by HungryProton** [comment]
I can imagine a few tricky edge cases if the cache works in a per-node basis, especially when saving Scatter nodes in sub scenes and reusing them in different places.

I think this could work nicely:

+ Create a `ScatterCache` node, there should be at most one per scene.
+ This node scans for other scatter node in the scene, excluding instanced scenes. 
  - This step doesn't have to be automatic for now, a button `Rebuild cache` should work fine at first.
+ If the transforms are ready, stores them in a custom resource.
+ That cache resource should be stored either:
  - Next to the current scene.
  - Or in a dedicated cache folder in the add-on folder.
  - The user should be able to pick a path, I think. Not sure if this should be a project setting or set on each individual cache.
+ ScatterItems node could have a `ignore_cache` option to get more granular control.
+ When the game (or the editor) starts, if there's a cache node in the scene, it should restore the transforms from the cache file.
  - Which means, `force_rebuild_on_load` should be automatically disabled by the ScatterCache node.


That sounds like a lot of work, though. If you really need this feature feel free to go ahead, otherwise I don't mind doing it myself, but I don't know when I'll get to start working on this.

**Comment by HungryProton** [comment]
I'll start working on this issue this weekend, unless you already got something on your side, or new ideas?
I'll try to implement the system described in my previous comment.

**Comment by TokisanGames** [comment]
If a cache is too much work, how about just storing an array in a binary .res file that the user can manually specify? Then on every save it writes to the res if it exists, or it saves in the scene if not.

Also if a cache, can it be included in the export like a res can?

The data could be an array of locations, or it could be a splat map image that is provided to the GPU particle shader.

**Comment by HungryProton** [comment]
The cache is going to be exactly that: a Godot resource saved on disk, so yes it can be exported like any other resource.
The shader cache node is only there to do all the saving / reloading on other scatter nodes in the scene.

**Comment by HungryProton** [comment]
Implemented in #120 
I'll close the issue once I find a good way to automatically rebuild the cache when closing or saving a scene. For now, users have to manually click on the Rebuild button in the cache node.

**Comment by TokisanGames** [comment]
Great, we'll test it this week. 

So in order to save changes we must manually click rebuild? Why not grab NOTIFICATION_EDITOR_PRE_SAVE in _notify inside the editor plugin and trigger it there if any children nodes have changed?

https://docs.godotengine.org/en/stable/classes/class_node.html#constants

**Comment by HungryProton** [comment]
Good idea, it's done in 485cd4f6fd9103dc7f6c239841d3399ba6236666
